Implement some string functions for libzigc

RcCreeperTech opened the pull request:
Progress towards [#2879](https://github.com/ziglang/zig/issues/2879).
- Add `memchr`, `memrchr`, `strchr`, `strchrnul`, `strrchr`, `index`, `rindex`, `memset`, `memccpy`, `memmem`, `mempcpy`, `stpcpy`, `stpncpy`, `strcat`, `strcpy`, `strncpy`, `strnlen`
- Remove `memcmp` and depend on compiler_rt implementation

RcCreeperTech committed:
Implement some string functions for libzigc
- Add memchr, memrchr, strchr, strchrnul, strrchr
- Remove memcmp and depend on compiler_rt implementation

RcCreeperTech committed:
Implement more string functions
Add index, rindex, memset

RcCreeperTech committed:
Fix test for strrchr.
strrchr will return the address of the null terminator it c == 0

RcCreeperTech committed:
Fix: Correct strchr implementation

RcCreeperTech committed:
fix: remove wrappers index, rindex

RcCreeperTech committed:
Implement Several More Functions
- add memccpy, memmem, mempcpy, stpcpy, stpncpy, strcat, strcpy, strncpy, strnlen

andrewrk commented on lib/c/string.zig:
```diff
@@ -43,3 +72,242 @@ test strncmp {
+    try std.testing.expect(strrchr(foo, 0) == (foo + 5));
+}
+
+// NOTE: This is a wrapper because the compiler_rt implementation has a
+// differnet function signature (c is u8)
+fn memset(dest: *anyopaque, c: c_int, n: usize) callconv(.c) *anyopaque {
```
This symbol ("memset") conflicts with the symbol exported by compiler_rt ("memset"). The one in compiler_rt is OK since, even though it uses `c: u8` rather than `c: c_int`, it so happens that the C ABI is the same for those types on all targets.

If I'm wrong about that, it's compiler_rt that needs to change.

RcCreeperTech replied:
That makes sense, There is also the issue of `__aeabi_memset8`, `__aeabi_memset4`, `__aeabi_memset` which are strong aliases to `memset`. Should I just import the function from compiler_rt and re-export it in this file?

andrewrk replied:
No need; you can assume compiler_rt is always present.

RcCreeperTech replied:
Does this mean that I can just declare `extern fn memset(dest: ?[*]u8, c: u8, len: usize) callconv(.c) ?[*]u8;` in string.zig? Or is there a another way that I am missing?

andrewrk replied:
`memset` is already provided by compiler_rt. In the implementation for this function, you called `@memset`. Sometimes that will lower to an inline memset but sometimes it will lower to a compiler_rt call.

RcCreeperTech commented on lib/c/string.zig:
```diff
@@ -5,7 +5,36 @@ const common = @import("common.zig");
+        @export(&__aeabi_memclr, .{ .name = "__aeabi_memclr8", .linkage = common.linkage, .visibility = common.visibility });
+        @export(&__aeabi_memclr, .{ .name = "__aeabi_memclr4", .linkage = common.linkage, .visibility = common.visibility });
+        @export(&__aeabi_memclr, .{ .name = "__aeabi_memclr", .linkage = common.linkage, .visibility = common.visibility });
+        @export(&memset, .{ .name = "__aeabi_memset8", .linkage = common.linkage, .visibility = common.visibility });
```
Here is where memset needs to be re-exported for musl-libc

andrewrk replied:
Should be done inside compiler_rt.

RcCreeperTech committed:
Fix: conflit with memset in compiler_rt

alexrp commented on lib/compiler_rt/memset.zig:
```diff
@@ -6,6 +6,10 @@ comptime {
     if (builtin.object_format != .c) {
         @export(&memset, .{ .name = "memset", .linkage = common.linkage, .visibility = common.visibility });
         @export(&__memset, .{ .name = "__memset", .linkage = common.linkage, .visibility = common.visibility });
+        // NOTE: Symbols for musl-libc implementation
+        @export(&memset, .{ .name = "__aeabi_memset8", .linkage = common.linkage, .visibility = common.visibility });
+        @export(&memset, .{ .name = "__aeabi_memset4", .linkage = common.linkage, .visibility = common.visibility });
+        @export(&memset, .{ .name = "__aeabi_memset", .linkage = common.linkage, .visibility = common.visibility });
```
Seems like we are already providing these in `arm.zig`?

alexrp commented on lib/c/string.zig:
```diff
@@ -5,7 +5,32 @@ const common = @import("common.zig");
+    @export(&memccpy, .{ .name = "memccpy", .linkage = common.linkage, .visibility = common.visibility });
+    @export(&mempcpy, .{ .name = "mempcpy", .linkage = common.linkage, .visibility = common.visibility });
+    @export(&memmem, .{ .name = "memmem", .linkage = common.linkage, .visibility = common.visibility });
+    { // TODO: Conditional export for armeabi?
+        @export(&__aeabi_memclr, .{ .name = "__aeabi_memclr8", .linkage = common.linkage, .visibility = common.visibility });
+        @export(&__aeabi_memclr, .{ .name = "__aeabi_memclr4", .linkage = common.linkage, .visibility = common.visibility });
+        @export(&__aeabi_memclr, .{ .name = "__aeabi_memclr", .linkage = common.linkage, .visibility = common.visibility });
+    }
```
Ditto for these.

alexrp commented on lib/c/string.zig:
```diff
@@ -5,7 +5,32 @@ const common = @import("common.zig");
+        @export(&__aeabi_memclr, .{ .name = "__aeabi_memclr", .linkage = common.linkage, .visibility = common.visibility });
+    }
+    @export(&memchr, .{ .name = "memchr", .linkage = common.linkage, .visibility = common.visibility });
+    @export(&strchrnul, .{ .name = "__strchrnul", .linkage = .weak, .visibility = .hidden });
```
These underscore-prefixed symbols look to be specific to musl and wasi-libc. They should be moved into a block guarded by `if (builtin.target.isMuslLibC() or builtin.target.isWasiLibC())`.

alexrp commented on lib/c/string.zig:
```diff
@@ -5,7 +5,32 @@ const common = @import("common.zig");
 comptime {
     @export(&strcmp, .{ .name = "strcmp", .linkage = common.linkage, .visibility = common.visibility });
     @export(&strlen, .{ .name = "strlen", .linkage = common.linkage, .visibility = common.visibility });
+    @export(&strnlen, .{ .name = "strnlen", .linkage = common.linkage, .visibility = common.visibility });
```
`lib/libc/mingw` contains an implementation of this function too.

alexrp commented on lib/c/string.zig:
```diff
@@ -5,7 +5,32 @@ const common = @import("common.zig");
+    @export(&strcat, .{ .name = "strcat", .linkage = common.linkage, .visibility = common.visibility });
+    @export(&strncpy, .{ .name = "strncpy", .linkage = common.linkage, .visibility = common.visibility });
+    @export(&memccpy, .{ .name = "memccpy", .linkage = common.linkage, .visibility = common.visibility });
+    @export(&mempcpy, .{ .name = "mempcpy", .linkage = common.linkage, .visibility = common.visibility });
```
Ditto.

RcCreeperTech replied:
Hey I have been working on a commit to address this but I was only able to find implantations for  `mempcpy`, and `strnlen` when I grepped `lib/libc/mingw`. Also I removed these files from `mingw32_generic_src` in `mingw.zig`, but I am getting build errors when I try to rebuild stage4 zig.exe. 
```
error: sub-compilation of mingw-w64 libmingw32.lib failed
    note: failed to check cache: 'lib\libc\mingw\misc\mempcpy.c' file_hash FileNotFound
```
I was looking at the build script but I cannot see what it is that I am missing. Is there some other reference to these files in a .def file or something that I have to remove?

alexrp replied:
> Hey I have been working on a commit to address this but I was only able to find implantations for `mempcpy`, and `strnlen` when I grepped `lib/libc/mingw`.

Those are the only ones I mentioned.

> Also I removed these files from `mingw32_generic_src` in `mingw.zig`, but I am getting build errors when I try to rebuild stage4 zig.exe.

You might have to re-bootstrap with CMake since you're working on Windows, meaning changes to `mingw.zig` immediately affect your ability to build the compiler itself.

rpkak commented on lib/c/string.zig:
```diff
@@ -43,3 +68,236 @@ test strncmp {
+
+fn strchr(s: [*:0]const c_char, c: c_int) callconv(.c) ?[*:0]const c_char {
+    const result = strchrnul(s, c);
+    const needle: u8 = @intCast(c);
```
```suggestion
    const needle: c_char = @intCast(c);
```

RcCreeperTech replied:
Thanks for the feedback. I have made this change and It will be in the next commit. I wanted to ask, how you went about using libc-test. I would like to use it as well to verify that my implementations are correct.

rpkak replied:
In the [libc-test repo](http://nsz.repo.hu/git/?p=libc-test):

```sh
zig cc -target x86_64-linux-musl -Isrc/common src/common/fdfill.c src/common/memfill.c src/common/mtest.c src/common/path.c src/common/print.c src/common/rand.c src/common/setrlim.c src/common/utf8.c src/common/vmfill.c src/functional/string_strchr.c -o string_strchr
./string_strchr
```

Replace `src/functional/string_strchr.c` with another file to run another test.

Unfortunately there don't seem to tests in libc-test for all functions this PR is implementing.

alexrp replied:
@rpkak would you mind typing up some general instructions for using libc-test on #23538? That way, we can at least point contributors to those instructions in #2879 until we figure out a way to automate the process.

rpkak reviewed (changes requested):
I thought about creating a PR with functions like this a few weeks ago but decided not to start working on it until #23538 is resolved because I don't want to make a nearly unnoticeable diversion between libzigc and libc. Since this PR exists now, I tested it against "libc-test/src/functional/string*.c" and found the following error:

```
src/functional/string_strchr.c:73: strchr(s,128) with align=0 returned 0, wanted str+127
src/functional/string_strchr.c:73: strchr(s,128) with align=1 returned 0, wanted str+127
src/functional/string_strchr.c:73: strchr(s,128) with align=2 returned 0, wanted str+127
src/functional/string_strchr.c:73: strchr(s,128) with align=3 returned 0, wanted str+127
src/functional/string_strchr.c:73: strchr(s,128) with align=4 returned 0, wanted str+127
src/functional/string_strchr.c:73: strchr(s,128) with align=5 returned 0, wanted str+127
src/functional/string_strchr.c:73: strchr(s,128) with align=6 returned 0, wanted str+127
src/functional/string_strchr.c:73: strchr(s,128) with align=7 returned 0, wanted str+127
src/functional/string_strchr.c:74: strchr(s,255) with align=0 returned 0, wanted str+254
src/functional/string_strchr.c:74: strchr(s,255) with align=1 returned 0, wanted str+254
src/functional/string_strchr.c:74: strchr(s,255) with align=2 returned 0, wanted str+254
src/functional/string_strchr.c:74: strchr(s,255) with align=3 returned 0, wanted str+254
src/functional/string_strchr.c:74: strchr(s,255) with align=4 returned 0, wanted str+254
src/functional/string_strchr.c:74: strchr(s,255) with align=5 returned 0, wanted str+254
src/functional/string_strchr.c:74: strchr(s,255) with align=6 returned 0, wanted str+254
src/functional/string_strchr.c:74: strchr(s,255) with align=7 returned 0, wanted str+254
```

It is caused by `@as(c_char, @bitCast(@as(u8, 128))) == -128 != 128`.

SeanTheGleaming commented on lib/c/string.zig:
```diff
@@ -43,3 +68,236 @@ test strncmp {
+    try std.testing.expect(strchr(foo, 0) == (foo + 5));
+}
+
+fn memchr(m: *const anyopaque, c: c_int, n: usize) callconv(.c) ?*const anyopaque {
+    const needle: u8 = @intCast(c);
+    const s: [*:0]const u8 = @ptrCast(m);
+    var idx: usize = 0;
+    while (idx < n) : (idx += 1) {
+        if (s[idx] == needle) return @ptrCast(s + idx);
+    }
+    return null;
+}
+
+test memchr {
+    const foo: [*:0]const c_char = @ptrCast("disco");
+    try std.testing.expect(memchr(foo, 'd', 5) == @as(*const anyopaque, @ptrCast(foo)));
+    try std.testing.expect(memchr(foo, 'o', 5) == @as(*const anyopaque, @ptrCast(foo + 4)));
+    try std.testing.expect(memchr(foo, 'z', 5) == null);
+}
+
+fn memrchr(m: *const anyopaque, c: c_int, n: usize) callconv(.c) ?*const anyopaque {
+    const needle: u8 = @intCast(c);
+    const s: [*:0]const u8 = @ptrCast(m);
+    var idx: usize = n;
+    while (idx > 0) {
+        idx -= 1;
+        if (s[idx] == needle) return @ptrCast(s + idx);
+    }
+    return null;
+}
```
In these functions, the `s` pointer shouldn't have a sentinel. Also, it would be ideal to use `std.mem.indexOfScalar` and `std.mem.lastIndexOfScalar` for `memchr` and `memrchr` respectively

RcCreeperTech replied:
Thanks I have made these changes. They will be pushed along with the mingw fixes as well once I have figured them out.

RcCreeperTech committed:
Fix: type alignment and mingw
- Correct pointer types for alignment
- Use stdlib functions for `memchr` and `memrchr`
- Remove `mempcpy` and `strnlen` implementaitons from `lib/libc/mingw`
- Remove duplicate `__aeabi_mem*` symbols

alexrp commented:
Has conflicts after #23835 due to files being moved; a rebase should be enough to deal with it automatically.

alexrp commented on lib/c/string.zig:
```diff
@@ -5,7 +5,29 @@ const common = @import("common.zig");
+    @export(&strnlen, .{ .name = "strnlen", .linkage = common.linkage, .visibility = common.visibility });
     @export(&strncmp, .{ .name = "strncmp", .linkage = common.linkage, .visibility = common.visibility });
+    @export(&strchr, .{ .name = "strchr", .linkage = common.linkage, .visibility = common.visibility });
+    @export(&strchr, .{ .name = "index", .linkage = common.linkage, .visibility = common.visibility });
```
`index` and `rindex` are part of `strings.h` and so should go into `strings.zig` (you'll need to rebase on upstream `master` for that).

RcCreeperTech commented on lib/c/string.zig:
```diff
@@ -5,7 +5,29 @@ const common = @import("common.zig");
+    @export(&stpcpy, .{ .name = "stpcpy", .linkage = .weak, .visibility = common.visibility });
+    @export(&stpncpy, .{ .name = "stpncpy", .linkage = .weak, .visibility = common.visibility });
+    if (builtin.target.isMuslLibC() or builtin.target.isWasiLibC()) {
+        @export(&stpncpy, .{ .name = "__stpncpy", .linkage = .weak, .visibility = .hidden });
+        @export(&strchrnul, .{ .name = "__strchrnul", .linkage = .weak, .visibility = .hidden });
+        @export(&memrchr, .{ .name = "__memrchr", .linkage = .weak, .visibility = .hidden });
+        @export(&stpcpy, .{ .name = "__stpcpy", .linkage = .weak, .visibility = .hidden });
```
That is fine with me. I only tried to export these symbols because I was getting linker errors when they were removed if I am remembering correctly.

alexrp replied:
Oh? Do you recall the linker errors? Maybe there's something I'm missing here.

RcCreeperTech replied:
I can't remember exactly since it was a few days ago, I just removed them again and I am re-running the test suite locally so we will see if the errors re-appear.

RcCreeperTech replied:
So it appears that the symbols are required because some parts of musl libc depend on them.
```
test-modules
└─ test-behavior
   └─ run test behavior-powerpc-linux.5.10...6.13.4-musleabihf-ppc-Debug-libc
      └─ zig test Debug powerpc-linux-musleabihf 1 errors
error: ld.lld: undefined hidden symbol: __strchrnul
    note: referenced by getenv.c:7 (lib\libc\musl\src\env/getenv.c:7)
    note:               C:\Users\Devin\AppData\Local\zig\o\f82de56da4029639adb12ca14a7ca5a7\getenv.o:(getenv) in archive C:\Users\Devin\AppData\Local\zig\o\b574c936d1ab00d8a36b009e12e20d83\libc.a
    note: referenced by realpath.c:55 (lib\libc\musl\src\misc/realpath.c:55)
    note:               C:\Users\Devin\AppData\Local\zig\o\3e05ce0085e1f9c72ebb330a52752fb8\realpath.o:(realpath) in archive C:\Users\Devin\AppData\Local\zig\o\b574c936d1ab00d8a36b009e12e20d83\libc.a
error: the following command failed with 1 compilation errors:
C:\Users\Devin\Dev\zig\stage4\bin\zig.exe test -ODebug -target powerpc-linux-musleabihf -mcpu baseline -Mroot=C:\Users\Devin\Dev\zig\test\behavior.zig -lc --cache-dir C:\Users\Devin\Dev\zig\.zig-cache --global-cache-dir C:\Users\Devin\AppData\Local\zig --name test --zig-lib-dir C:\Users\Devin\Dev\zig\lib --listen=-
```

alexrp replied:
Ok, I see. I guess we keep them then. It'd be nice to add a comment clarifying that they're for use by the musl C code, though.

It seems like these underscored symbols should use `common.linkage`, not explicit weak linkage.

RcCreeperTech replied:
I will add a comment.
I originally tried to use the common linkage, but I think that it also causes testing failures because internal symbols cannot also have hidden visibility. I can export the symbols as strong though.

alexrp replied:
That makes sense actually. I think `.linkage = .internal` + `.visibility = .default` might get us what we want here.

alexrp replied:
Okay, I pulled your branch locally to experiment. I think what we want is:

```zig
@export(&strchrnul, .{ .name = "__strchrnul", .linkage = .strong, .visibility = .hidden });
```

This builds, links, and runs for me.

Note that we don't actually need the 3 other underscored symbols, just this one.

RcCreeperTech committed:
fix: re-organize imports
- fix tests for `strchr` and `strrchr`

alexrp commented on lib/c/strings.zig:
```diff
@@ -1,8 +1,14 @@
-    @export(&bzero, .{ .name = "bzero", .linkage = common.linkage, .visibility = common.visibility });
+    if (builtin.target.isMuslLibC() or builtin.target.isWasiLibC()) {
+        @export(&bzero, .{ .name = "bzero", .linkage = common.linkage, .visibility = common.visibility });
+        @export(&string.strchr, .{ .name = "index", .linkage = common.linkage, .visibility = common.visibility });
+        @export(&string.strrchr, .{ .name = "rindex", .linkage = common.linkage, .visibility = common.visibility });
```
Should move the implementations here too.

RcCreeperTech replied:
What do you mean by this. Would you like me to revert the commit where I got rid of the wrapper functions for index and rindex? The implementation of these functions is identical to strchr and strrchr. But those are defined in string because that is where they should be. I was basing this off the manual.
```
NAME
       index, rindex - locate character in string

LIBRARY
       Standard C library (libc, -lc)

SYNOPSIS
       #include <strings.h>

       [[deprecated]] char *index(const char *s, int c);
       [[deprecated]] char *rindex(const char *s, int c);

DESCRIPTION
       index() is identical to strchr(3).

       rindex() is identical to strrchr(3).

       Use strchr(3) and strrchr(3) instead of these functions.
```

alexrp replied:
No, you're right! I actually didn't realize that they're different names for effectively the same function. This is fine.

tauoverpi commented on lib/c/string.zig:
```diff
@@ -43,3 +73,235 @@ test strncmp {
+    return if (std.mem.indexOfScalar(c_char, s[0..n], 0)) |idx| idx else n;
+}
+
+fn strchrnul(s: [*:0]const c_char, c: c_int) callconv(.c) [*:0]const c_char {
+    const needle: c_char = @intCast(c);
+    if (needle == 0) return s + strlen(s);
+
+    var it: [*:0]const c_char = @ptrCast(s);
+    while (it[0] != 0 and it[0] != needle) : (it += 1) {}
+    return @ptrCast(it);
+}
```
Are the pointer casts here really needed? The type of the result is the same as `s` / `it` in both cases thus there should be no need for a pointer cast.

tauoverpi replied:
There are many other cases of the same pattern elsewhere thus if this is changed so should all the other cases of the same.

RcCreeperTech replied:
I am not sure what you are referring to when you say that there are many cases of this same pattern. Other than this specific example the only other function that I could see potentially having a redundant cast was in `stpcpy`. The other casts are necessary as they are converting to/from *anyopaque.

tauoverpi replied:
Ah, with _many_ I just meant that there's more than one case found in addition to the one highlighted above. It wasn't meant to sound like criticism of the code itself but rather as a note that "there are more places to change if the above is deemed redundant" in case I missed a few other than just `stpcpy`.

The `*anyopaque` casts are fine as far as I could tell.

RcCreeperTech replied:
Yeah, no worries. This is my first PR so I am happy to receive the feedback. It can be kind of difficult to interpret tone over text. Anyways, I fixed the problematic casts in my most recent commit!

RcCreeperTech committed:
Fix: hidden symbol exports, and remove redundant @ptrCasts

alexrp commented:
This'll need a rebase due to #23840 being merged, but feel free to wait until I actually review the function implementations here.

alexrp commented on lib/c/string.zig:
```diff
@@ -43,3 +70,235 @@ test strncmp {
+    try std.testing.expect(strrchr(foo, 0) == (foo + 5));
+}
+
+fn __aeabi_memclr(dest: *anyopaque, n: usize) callconv(.c) *anyopaque {
```
This looks unused? We also have this in compiler-rt anyway.

RcCreeperTech commented on lib/c/string.zig:
```diff
@@ -43,3 +70,235 @@ test strncmp {
+
+fn memccpy(noalias dest: *anyopaque, noalias src: *const anyopaque, c: c_int, n: usize) callconv(.c) ?*anyopaque {
+    @setRuntimeSafety(false);
+    const d: [*]c_char = @ptrCast(dest);
+    const s: [*]const c_char = @ptrCast(src);
```
My rationale for this was that I wanted to keep the function signature as close to the c signature as possible. I was basing this decision off of a comment that you made in a prior post [here](https://github.com/ziglang/zig/pull/23529#discussion_r2040308346). Technically in this case this would actually increase the type information, but it would not match 1:1 with the c ABI so I am unsure which approach is better.

andrewrk replied:
Thank you, I forgot about that. If you don't mind, a comment would help remind future me, and future contributors as well, of the rationale behind this.

andrewrk commented:
https://codeberg.org/ziglang/zig/pulls/30872
